File: md5.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <openssl/md5.h>
#include <omp.h>
#include <algorithm>
#include <set>
#include <cmath>
#include <cstring>

using namespace std;

struct Target {
    string filename;
    string login;
    string salt;
    unsigned char hash[16];
    bool cracked = false;
    string password;
    string category;
};

// Base64 decoding to binary hash
vector<unsigned char> base64_decode(const string& in) {
    static const string b64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    vector<int> T(256, -1);
    for (int i = 0; i < 64; i++) T[b64[i]] = i;
    vector<unsigned char> out;
    int val = 0, valb = -8;
    for (unsigned char c : in) {
        if (T[c] == -1) break;
        val = (val << 6) + T[c];
        valb += 6;
        if (valb >= 0) {
            out.push_back((unsigned char)((val >> valb) & 0xFF));
            valb -= 8;
        }
    }
    return out;
}

// Global logging to console and file
void log_cracked(Target& t, string pass, string cat) {
    t.cracked = true;
    t.password = pass;
    t.category = cat;
    string msg = "  [+] CRACKED: " + t.filename + " | " + t.login + " | " + pass + " (" + cat + ")";
    cout << msg << endl;
    ofstream out("results.txt", ios::app);
    out << msg << endl;
}

// --- Category 1: Dictionary Attack (Full Scan) ---
void run_category1_full(vector<Target>& targets) {
    cout << "[Category 1] Starting full Dictionary Attack on all hashes..." << endl;

    vector<string> base_names = {
        "alexander", "alexandra", "adam", "adela", "adrian", "adriana", "alica", "alena", "andrea", "andrej",
        "anna", "anton", "barbora", "beata", "benjamin", "bozena", "branislav", "branko", "cyril", "daniela",
        "daniel", "darina", "david", "denisa", "dominika", "dominik", "dusan", "edita", "eduard", "elena",
        "emil", "emilia", "erik", "erika", "eva", "filip", "frantisek", "gabriel", "gabriela", "hana",
        "helena", "henrich", "ivan", "ivana", "iveta", "jakub", "jan", "janka", "jarmila", "jaroslav",
        "jozef", "julia", "julius", "juraj", "kamil", "karol", "katarina", "klara", "klaudia", "kristina",
        "ladislav", "laura", "lenka", "lucia", "ludmila", "lukas", "magdalena", "marek", "marian", "maria",
        "marta", "martin", "martina", "matej", "matus", "michal", "michaela", "milan", "milos", "miroslav",
        "monika", "natalia", "nikola", "nina", "olga", "oliver", "ondrej", "oskar", "patrik", "patricia",
        "pavol", "petra", "peter", "radovan", "radoslav", "rastislav", "renata", "richard", "robert",
        "roman", "samuel", "silvia", "simona", "simon", "stanislav", "stefan", "tamara", "tatiana",
        "terezia", "tibor", "timotej", "tomas", "veronika", "viera", "viktoria", "viktor", "viliam", "vladimir",
        "zdeno", "zora", "zuzana", "misko", "ferko", "janko", "katka", "zuzka", "evka", "lucka", "petko",
        "samko", "matko", "jozko", "matusko", "kubko"
    };

    vector<string> suffixes = {"ka", "ko", "icka", "ik", "uska", "ulo", "inka"};
    set<string> temp_dict;

    for (string name : base_names) {
        temp_dict.insert(name);
        string stem3 = name.substr(0, min((int)name.length(), 3));
        string stem4 = name.substr(0, min((int)name.length(), 4));
        for (string sfx : suffixes) {
            temp_dict.insert(stem3 + sfx);
            temp_dict.insert(stem4 + sfx);
        }
    }

    vector<string> dictionary;
    for (string word : temp_dict) {
        dictionary.push_back(word);
        for (int i = 0; i < (int)word.length(); i++) {
            string variant = word;
            variant[i] = toupper(variant[i]);
            dictionary.push_back(variant);
        }
    }

    double start = omp_get_wtime();
    int found_count = 0;

    #pragma omp parallel for schedule(dynamic) reduction(+:found_count)
    for (int i = 0; i < (int)targets.size(); i++) {
        for (const string& cand : dictionary) {
            MD5_CTX ctx;
            unsigned char digest[16];
            MD5_Init(&ctx);
            MD5_Update(&ctx, cand.c_str(), cand.length());
            MD5_Update(&ctx, targets[i].salt.c_str(), targets[i].salt.length());
            MD5_Final(digest, &ctx);

            if (memcmp(digest, targets[i].hash, 16) == 0) {
                #pragma omp critical
                {
                    if (!targets[i].cracked) {
                        log_cracked(targets[i], cand, "Category 1");
                        found_count++;
                    }
                }
                break;
            }
        }
    }
    double duration = omp_get_wtime() - start;
    string end_msg = "[Category 1] Found: " + to_string(found_count) + " passwords. Time: " + to_string(duration) + " s.";
    cout << end_msg << endl;
    ofstream out("results.txt", ios::app);
    out << end_msg << endl;
}

// --- Brute Force (Category 2 & 3) ---
void run_brute_force(vector<Target>& targets, int length, const string& charset, int cat_num) {
    set<string> files_needed = {"shadow1.txt", "shadow2.txt", "shadow3.txt", "shadow4.txt"};
    set<string> satisfied_files;
    for (auto& t : targets) {
        if (t.cracked && (t.category == "Category " + to_string(cat_num))) satisfied_files.insert(t.filename);
    }

    if (satisfied_files.size() >= 4) {
        cout << "[Category " << cat_num << "] Objective met, skipping length " << length << endl;
        return;
    }

    long long total = (long long)pow(charset.length(), length);
    cout << "[Category " << cat_num << "] L:" << length << " | " << total << " combinations..." << endl;

    bool stop_early = false;
    double start_time = omp_get_wtime();

    #pragma omp parallel
    {
        MD5_CTX ctx_base;
        #pragma omp for schedule(dynamic, 2000)
        for (long long i = 0; i < total; i++) {
            if (stop_early) continue;

            char pass[12];
            long long temp = i;
            for (int l = 0; l < length; l++) {
                pass[l] = charset[temp % charset.length()];
                temp /= charset.length();
            }
            pass[length] = '\0';

            for (int j = 0; j < (int)targets.size(); j++) {
                if (targets[j].cracked) continue;

                unsigned char digest[16];
                MD5_Init(&ctx_base);
                MD5_Update(&ctx_base, pass, length);
                MD5_Update(&ctx_base, targets[j].salt.c_str(), targets[j].salt.length());
                MD5_Final(digest, &ctx_base);

                if (memcmp(digest, targets[j].hash, 16) == 0) {
                    #pragma omp critical
                    {
                        if (!targets[j].cracked) {
                            log_cracked(targets[j], pass, "Category " + to_string(cat_num));
                            satisfied_files.insert(targets[j].filename);
                            if (satisfied_files.size() >= 4) stop_early = true;
                        }
                    }
                }
            }
        }
    }
    double duration = omp_get_wtime() - start_time;
    string time_msg = "  Phase [Category " + to_string(cat_num) + " L:" + to_string(length) + "] took: " + to_string(duration) + " s.";
    cout << time_msg << endl;
    ofstream out("results.txt", ios::app);
    out << time_msg << endl;
}

int main() {
    ofstream out("results.txt");
    out << "--- ATTACK STARTED: " << omp_get_wtime() << " ---" << endl;
    out.close();

    vector<Target> targets;
    vector<string> files = {"shadow1.txt", "shadow2.txt", "shadow3.txt", "shadow4.txt"};
    for (string fname : files) {
        ifstream f(fname);
        string line;
        while (getline(f, line)) {
            stringstream ss(line);
            string u, s, h;
            if (getline(ss, u, ':') && getline(ss, s, ':') && getline(ss, h, ':')) {
                Target t; t.filename = fname; t.login = u; t.salt = s;
                vector<unsigned char> dec = base64_decode(h);
                if (dec.size() == 16) { memcpy(t.hash, dec.data(), 16); targets.push_back(t); }
            }
        }
    }

    if (targets.empty()) return 1;
    cout << "Loaded " << targets.size() << " hashes." << endl;

    // 1. Dictionary Attack
    run_category1_full(targets);

    // 2. Mixed Alphanumeric (Category 3)
    string charset_mix = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    run_brute_force(targets, 4, charset_mix, 3);

    // 3. Lowercase (Category 2)
    string charset_low = "abcdefghijklmnopqrstuvwxyz";
    run_brute_force(targets, 6, charset_low, 2);

    cout << "\n--- FINISHED. Check results.txt ---" << endl;
    return 0;
}

File: md5.py
#!/usr/bin/env python3
import hashlib
import base64
import string
import itertools
import os
import time
import multiprocessing

OUTPUT_FILE = "vysledky.txt"
HEARTBEAT_INTERVAL = 60

def worker_optimized_generator(args):
    prefix, suffix_len, charset, targets_bin = args
    found = []
    prefix_bytes = prefix.encode('utf-8')
    base_m = hashlib.md5(prefix_bytes)

    for suffix_tuple in itertools.product(charset, repeat=suffix_len):
        suffix_str = "".join(suffix_tuple)
        suffix_bytes = suffix_str.encode('utf-8')
        m_pass = base_m.copy()
        m_pass.update(suffix_bytes)

        for salt_bytes, target_hash_bytes, login, fname in targets_bin:
            m_final = m_pass.copy()
            m_final.update(salt_bytes)
            if m_final.digest() == target_hash_bytes:
                found.append((fname, login, prefix + suffix_str))
    return found

def log_result(message):
    print(message)
    with open(OUTPUT_FILE, "a", encoding="utf-8") as f:
        f.write(message + "\n")

def load_shadow_files(directory):
    targets = []
    files = sorted([f for f in os.listdir(directory) if f.startswith("shadow") and f.endswith(".txt")])
    for filename in files:
        filepath = os.path.join(directory, filename)
        with open(filepath, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if not line or ':' not in line: continue
                parts = line.split(':')
                if len(parts) == 3:
                    login, salt, pwhash = parts
                    targets.append({
                        'file': filename, 'login': login,
                        'salt_bytes': salt.encode('utf-8'),
                        'hash_bytes': base64.b64decode(pwhash),
                        'cracked': None, 'cats_found': set()
                    })
    return targets

def get_category1_candidates():
    base_names = [
        "alexander", "alexandra", "adam", "adela", "adrian", "adriana", "alica", "alenka", "alena", "andrea", "andrej",
        "anna", "anezka", "anton", "barbora", "beata", "benjamin", "berta", "blanka", "bohumil", "bohumir", "bozena",
        "branislav", "branko", "brigita", "cyril", "dagmar", "dalibor", "dana", "daniela", "daniel", "darina", "david",
        "denisa", "denis", "dezider", "diana", "dionyz", "dobroslav", "dominika", "dominik", "drahomir", "dusan", "dusana",
        "edita", "eduard", "elena", "eleonora", "elizabeth", "ela", "emil", "emilia", "erik", "erika", "estera", "eva",
        "evzen", "filip", "frantisek", "frantiska", "gabriel", "gabriela", "gaspar", "gejza", "gerda", "gizela", "gustav",
        "hana", "hedviga", "helena", "henrich", "hilda", "hubert", "hugo", "ida", "ignac", "imrich", "irena", "ivan",
        "ivana", "iveta", "ivica", "izabela", "jakub", "jan", "janka", "jarmila", "jaroslav", "jaroslava", "jela",
        "jozef", "jozefina", "judita", "julia", "julius", "juraj", "justina", "kamila", "kamil", "karol", "karolina",
        "katarina", "kazimir", "klara", "klaudia", "klement", "koloman", "konstantin", "kornel", "kristian", "kristina",
        "krystof", "kveta", "ladislav", "laura", "lea", "lenka", "leonard", "leopold", "libusa", "linda", "livia", "lucia",
        "ludmila", "ludovit", "lukas", "lydia", "magdalena", "malvina", "marcela", "marek", "margita", "marian", "maria",
        "marta", "martin", "martina", "matej", "matus", "maximilian", "metod", "michaela", "michal", "mikulas", "milada",
        "milan", "miloslav", "milos", "miroslav", "miroslava", "monika", "natalia", "nela", "nikola", "nina", "nora",
        "norbert", "oldrich", "olga", "oliver", "ondrej", "oskar", "oto", "pankrac", "patrik", "patricia", "pavol",
        "pavel", "petra", "peter", "petronela", "pravoslav", "prokop", "radovan", "radoslav", "rastislav", "rebecca",
        "regina", "renata", "rene", "richard", "robert", "roman", "romana", "rozalia", "rudolf", "rut", "sabina",
        "samuel", "servac", "severin", "silvester", "silvia", "simona", "simon", "slavka", "slavomir", "sona",
        "stanislav", "stela", "svatopluk", "svetlana", "simon", "span", "stefan", "stefania", "tamara", "tatiana",
        "terezia", "tibor", "timotej", "tomas", "ulrika", "urban", "ursula", "vaclav", "valentin", "valeria", "vanda",
        "vavrinec", "vendelin", "veronika", "viera", "viktoria", "viktor", "viliam", "vincent", "viola", "vit",
        "vladimir", "vladislav", "vlasta", "vojtech", "xenia", "zoltan", "zora", "zuzana", "zigmund", "zofia"
    ]

    suffixes = ["ka", "ko", "ko", "icka", "ik", "uska", "ko", "ulo", "o", "usa", "inka"]

    all_words = set()
    for name in base_names:
        name = name.lower()
        all_words.add(name)

        stem = name[:3] if len(name) > 3 else name
        stem4 = name[:4] if len(name) > 4 else name

        for sfx in suffixes:
            all_words.add(stem + sfx)
            all_words.add(stem4 + sfx)

        all_words.update([
            name + "ko", name + "ka", name + "ik",
            "misko", "ferko", "janko", "katka", "zuzka", "evka", "lucka", "petko",
            "samko", "matko", "jozko", "matusko", "kubko", "petik", "janik"
        ])

    final_candidates = set()
    for word in all_words:
        clean_word = word.replace('a','a').replace('c','c').replace('d','d').replace('e','e').replace('i','i').replace('l','l').replace('n','n').replace('o','o').replace('r','r').replace('s','s').replace('t','t').replace('u','u').replace('y','y').replace('z','z')

        for w in [word, clean_word]:
            final_candidates.add(w.lower())
            for i in range(len(w)):
                variant = w[:i] + w[i].upper() + w[i+1:]
                final_candidates.add(variant)

    return list(final_candidates)

def check_all_files_satisfied(targets, category_label):
    all_files = set(t['file'] for t in targets)
    files_with_cat = set(t['file'] for t in targets if category_label in t['cats_found'])
    return all_files.issubset(files_with_cat)

def run_dictionary_attack(targets, candidates):
    log_result(f"\n[Category 1] Dictionary Attack...")
    start = time.time()
    category_label = "Category 1"
    for t in targets:
        for cand in candidates:
            m = hashlib.md5(cand.encode('utf-8'))
            m.update(t['salt_bytes'])
            if m.digest() == t['hash_bytes']:
                t['cracked'] = cand
                t['cats_found'].add(category_label)
                log_result(f"  [+] CRACKED: {t['file']} | {t['login']} | {cand}")
                break
    log_result(f"Phase Category 1 duration: {time.time()-start:.2f}s")

def run_generic_brute_force(targets, length, charset, category_label):
    if check_all_files_satisfied(targets, category_label):
        print(f"\n[{category_label}] Condition (1 password/file) already met. Skipping length {length}.")
        return

    active_targets_bin = [
        (t['salt_bytes'], t['hash_bytes'], t['login'], t['file'])
        for t in targets if not t['cracked']
    ]

    if not active_targets_bin:
        log_result(f"\n[{category_label}] No remaining hashes.")
        return

    print(f"\n[{category_label}] Brute force: length {length}...")
    cores = multiprocessing.cpu_count()
    pool = multiprocessing.Pool(processes=cores)
    tasks = [(char, length - 1, charset, active_targets_bin) for char in charset]

    start_time = time.time()
    last_heartbeat = start_time
    stop_early = False

    try:
        iterator = pool.imap_unordered(worker_optimized_generator, tasks)
        while not stop_early:
            try:
                result_list = iterator.next(timeout=1)
                for fname, login, password in result_list:
                    for t in targets:
                        if t['login'] == login and t['file'] == fname and not t['cracked']:
                            t['cracked'] = password
                            t['cats_found'].add(category_label)
                            log_result(f"  [+] CRACKED {category_label}: {fname} | {login} | {password}")

                    if check_all_files_satisfied(targets, category_label):
                        log_result(f"  [!] Condition for {category_label} met for all files. Ending length {length} early.")
                        stop_early = True
                        break
            except multiprocessing.TimeoutError:
                now = time.time()
                if now - last_heartbeat >= HEARTBEAT_INTERVAL:
                    print(f"  ... [HEARTBEAT] {category_label} (L:{length}) running for {int((now-start_time)//60)}m")
                    last_heartbeat = now
            except StopIteration:
                break
    finally:
        pool.terminate()
        pool.join()
        log_result(f"Phase {category_label} (L:{length}) duration: {time.time()-start_time:.2f}s")

def main():
    current_dir = os.path.dirname(os.path.abspath(__file__))
    with open(os.path.join(current_dir, OUTPUT_FILE), "w", encoding="utf-8") as f:
        f.write(f"--- ATTACK START: {time.ctime()} ---\n")

    targets = load_shadow_files(current_dir)

    run_dictionary_attack(targets, get_category1_candidates())

    charset_mix = string.ascii_letters + string.digits
    run_generic_brute_force(targets, 4, charset_mix, "Category 3")

    charset_low = string.ascii_lowercase
    run_generic_brute_force(targets, 6, charset_low, "Category 2")

if __name__ == "__main__":
    main()